processor/schemaprocessor/internal/translation/translation.go

```diff
@@ -0,0 +1,341 @@
+					return err
+				}
+				err = rev.logs.Rollback(log)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	in.SetSchemaUrl(t.schemaURL)
+	return nil
+}
+
+func (t *translator) ApplyScopeSpanChanges(ctx context.Context, scopeSpans ptrace.ScopeSpans, inSchemaURL string) error {
+	_, ver, err := GetFamilyAndVersion(inSchemaURL)
+	if err != nil {
+		return err
+	}
+	it, status := t.iterator(ctx, ver)
+	for rev, more := it(); more; rev, more = it() {
+		for i := 0; i < scopeSpans.Spans().Len(); i++ {
+			span := scopeSpans.Spans().At(i)
+			switch status {
+			case Update:
+				err = rev.all.Apply(span)
+				if err != nil {
+					return err
+				}
+				err = rev.spans.Apply(span)
+				if err != nil {
+					return err
+				}
+				for e := 0; e < span.Events().Len(); e++ {
+					event := span.Events().At(e)
+					err = rev.all.Apply(event)
+					if err != nil {
+						return err
+					}
+				}
+				if err = rev.spanEvents.Apply(span); err != nil {
+					return err
+				}
+			case Revert:
+				if err = rev.spanEvents.Rollback(span); err != nil {
+					return err
+				}
+				for e := 0; e < span.Events().Len(); e++ {
+					event := span.Events().At(e)
+					err = rev.all.Rollback(event)
+					if err != nil {
+						return err
+					}
+				}
+				err = rev.spans.Rollback(span)
+				if err != nil {
+					return err
+				}
+				err = rev.all.Rollback(span)
+				if err != nil {
+					return err
+				}
+			}
+		}
+		scopeSpans.SetSchemaUrl(t.schemaURL)
+	}
+	return nil
+}
+
+func (t *translator) ApplyScopeMetricChanges(ctx context.Context, in pmetric.ScopeMetrics, inSchemaURL string) error {
+	_, ver, err := GetFamilyAndVersion(inSchemaURL)
+	if err != nil {
+		return err
+	}
+	it, status := t.iterator(ctx, ver)
+	for rev, more := it(); more; rev, more = it() {
+		for i := 0; i < in.Metrics().Len(); i++ {
+			metric := in.Metrics().At(i)
+			switch status {
+			case Update:
+				if err = rev.all.Apply(metric); err != nil {
+					return err
+				}
+				if err = rev.metrics.Apply(metric); err != nil {
+					return err
+				}
+			case Revert:
+				if err = rev.metrics.Rollback(metric); err != nil {
+					return err
+				}
+				if err = rev.all.Rollback(metric); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	in.SetSchemaUrl(t.schemaURL)
+	return nil
+}
+
+// iterator abstractions the logic to perform an the migrations of, "From Version to Version".
+// The return values an iterator type and translation status that
+// should be compared against Revert, Update, NoChange
+// to determine what should be applied.
+// In the event that the ChangeSet has not yet been created (it is possible on a cold start)
+// then the iterator will wait til the update has been made
+//
+// Note: Once an iterator has been made, the passed context MUST cancel or run to completion
+//
+//	in order for the read lock to be released if either Revert or Upgrade has been returned.
+func (t *translator) iterator(ctx context.Context, from *Version) (iterator, int) {
```

Comment:
Do we need the extra complexity Context brings here? This is not a function that can asynchronously block on some external network call, it is a local execution with a reasonably low-bounded time. Do we need cancellation support here? What's wrong with letting all translation to run to the end and the caller to then cancel whatever it is that it was doing?